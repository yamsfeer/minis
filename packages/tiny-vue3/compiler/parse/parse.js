export const nodeTypes = {
  Root: Symbol('Root'),
  Element: Symbol('Element'),
  Attributes: Symbol('Attributes'),
  Interpolation: Symbol('Interpolation'),
  Expression: Symbol('Expression'),
  Text: Symbol('Text'),
}

const State = {
  DATA: Symbol('DATA'),
  RAWTEXT: Symbol('RAWTEXT'),
}

// tokenize 和 analyse 是 parse 的基本实现
// 以下是完整版 parse，使用递归下降算法
export default function parse(str) {
  const context = {
    source: str,
    state: State.DATA,
    consume(num) {
      if (typeof num === 'string') {
        num = num.length
      }
      context.source = context.source.slice(num)
    },
    consumeSpaces() {
      const match = /^[\t\r\n\f\s]+/.exec(context.source)
      match && context.consume(match[0].length)
    }
  }

  const ancestors = []
  const nodes = parseChildren(context, ancestors)

  return {
    type: nodeTypes.Root,
    children: nodes
  }
}

function isEnd(context, ancestors) {
  if (!context.source) {
    return true // source 处理完成
  }

  // 祖先中找到匹配的标签, ancestors 是一个栈
  for (let i = ancestors.length - 1; i >= 0; i--) {
    if (context.source.startsWith(`</${ancestors[i].tag}`)) {
      return true
    }
  }

  return false
}

function parseChildren(context, ancestors) {
  const nodes = []

  while (!isEnd(context, ancestors)) {
    let node
    if (context.state === State.DATA) {
      // 注意这里的 source 都要往 context 取最新值
      if (context.source[0] === '<' && isAlphabet(context.source[1])) {
        node = parseElement(context, ancestors) // 标签元素
      } else if (context.source.startsWith('{{')) {
        node = parseInterpolation(context) // 插值
      }
    }

    if (!node) {
      node = parseText(context)
    }

    nodes.push(node)
  }

  return nodes
}

function isAlphabet(char) {
  return /[a-z]/i.test(char)
}

function parseTag(context, type = 'start') {
  const { consume, consumeSpaces } = context
  // <div name="xxx"></div>
  const startTagReg = /^<([a-z][^[\t\r\n\f\s/>]*)/i
  const endTagReg = /^<\/([a-z][^[\t\r\n\f\s/>]*)/i

  const [match, tag] = type === 'start'
    // match: '<div', tag: 'div'
    ? startTagReg.exec(context.source)
    // match: '</div', tag: 'div'
    : endTagReg.exec(context.source)

  consume(match)
  consumeSpaces()

  // attributes
  const props = parseAttributes(context)

  const isSelfClosing = context.source.startsWith('/>')
  consume(isSelfClosing ? '/>' : '>')

  return {
    type: nodeTypes.Element,
    tag,
    props,
    children: [],
    isSelfClosing
  }
}

function parseAttributes(context) {
  const { consume, consumeSpaces } = context
  const props = []

  while (
    !context.source.startsWith('>') &&
    !context.source.startsWith('/>')
  ) {
    const [name] = /^[^\t\r\n\f\s/>][^\t\r\n\f\s/>=]*/.exec(context.source)

    consume(name)
    consumeSpaces()
    consume('=')
    consumeSpaces()

    let value = ''
    const DOUBLE_QUOTE = `"`
    const SINGLE_QUOTE = `'`
    const [quote] = context.source
    const isQuoted = quote === SINGLE_QUOTE || quote === DOUBLE_QUOTE

    if (isQuoted) {
      consume(quote)
      const endQuoteIndex = context.source.indexOf(quote)
      if (endQuoteIndex > -1) {
        value = context.source.slice(0, endQuoteIndex)
        consume(value)
        consume(quote)
      } else {
        console.error('引号不匹配')
      }
    } else {
      [value] = /^[^\t\r\n\f\s>]+/.exec(context.source)
      consume(value)
    }
    consumeSpaces()

    props.push({
      type: nodeTypes.Attributes,
      name,
      value
    })
  }

  return props
}

function parseElement(context, ancestors) {
  const element = parseTag(context)
  if (element.isSelfClosing)
    return element

  ancestors.push(element)
  // 每次 parseChildren 之前的空格换行怎么处理
  context.consumeSpaces()
  element.children = parseChildren(context, ancestors)
  context.consumeSpaces()
  ancestors.pop()

  if (context.source.startsWith(`</${element.tag}`)) {
    parseTag(context, 'end')
  } else {
    console.error(`${element.tag}缺少闭合标签`)
  }

  return element
}

function parseText(context) {
  let index = context.source.length // 有可能后面全是文本
  const quoteIndex = context.source.indexOf('<')
  const delimiterIndex = context.source.indexOf('{{')

  // 提取 '<' 之前的内容
  if (quoteIndex > -1) {
    index = quoteIndex
  }
  // 在 '<' 之前还有 '{{'，说明是文本中的插值，提取 '{{' 之前的内容
  if (delimiterIndex > -1 && delimiterIndex < index) {
    index = delimiterIndex
  }

  const content = context.source.slice(0, index)
  context.consume(content)

  return {
    type: nodeTypes.Text,
    content
  }
}

function parseInterpolation(context) {
  const { consume } = context

  consume('{{')
  const closeIndex = context.source.indexOf('}}')
  if (closeIndex < 0) {
    console.error('缺少结束定界符')
  }

  const content = context.source.slice(0, closeIndex)

  consume(content)
  consume('}}')

  return {
    type: nodeTypes.Interpolation,
    content: {
      type: nodeTypes.Expression,
      content
    }
  }
}
