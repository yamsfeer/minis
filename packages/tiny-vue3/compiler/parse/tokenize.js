const State = {
  data: Symbol('data'),

  tagOpen: Symbol('tagOpen'),
  tagName: Symbol('tagName'),

  text: Symbol('text'),

  tagEnd: Symbol('tagEnd'),
  tagEndName: Symbol('tagEndName'),
}

export const tokenTypes = {
  tagOpen: Symbol('tagOpen'),
  text: Symbol('text'),
  tagEnd: Symbol('tagEnd'),
}

const createTagOpen = (tagName) => ({ type: tokenTypes.tagOpen, name: tagName })
const createText = (text) => ({ type: tokenTypes.text, content: text })
const createTagEnd = (tagName) => ({ type: tokenTypes.tagEnd, name: tagName })

export default function tokenize(str) {
  const tokens = []
  let state = State.data
  let tmp = '' // 字符拼接
  let c

  while (str.length) {
    c = str[0] // 当前字符
    switch (state) {
      case State.data:
        if (c === '<') { state = State.tagOpen }
        else if (isAlphabet(c)) {
          state = State.text
          tmp += c
        }

        break
      case State.tagOpen:
        if (isAlphabet(c)) {
          state = State.tagName
          tmp += c
        }
        else if (c === '/') { state = State.tagEnd }

        break
      case State.tagName:
        if (isAlphabet(c)) { tmp += c }
        else if (c === '>') { // 取得一个 tagName
          state = State.data
          tokens.push(createTagOpen(tmp))
          tmp = ''
        }

        break
      case State.text:
        if (isAlphabet(c)) { tmp += c }
        else if (c === '<') {
          state = State.tagOpen
          tokens.push(createText(tmp))
          tmp = ''
        }

        break
      case State.tagEnd:
        if (isAlphabet(c)) {
          state = State.tagEndName
          tmp += c
        }
        break
      case State.tagEndName:
        if (isAlphabet(c)) { tmp += c }
        else if (c === '>') {
          state = State.data
          tokens.push(createTagEnd(tmp))
          tmp = ''
        }
        break

      default: break
    }
    str = str.slice(1) // consume
  }

  return tokens
}

function isAlphabet(char) {
  return /[a-zA-Z0-9]/.test(char)
}
