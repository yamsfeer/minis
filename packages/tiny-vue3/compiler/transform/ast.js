export const nodeTypes = {
  FunctionDecl: Symbol('FunctionDecl'), // 函数声明
  CallExpression: Symbol('CallExpression'), // 函数调用
  ReturnStatement: Symbol('ReturnStatement'), // 函数返回
  ArrayExpression: Symbol('ArrayExpression'), // 函数参数
  ObjectLiteral: Symbol('ObjectLiteral'), // 对象字面量
  StringLiteral: Symbol('StringLiteral'), // 字符串直接量
  Identifier: Symbol('Identifier'), // 标识符
  Interpolation: Symbol('Interpolation'), // 插值表达式
}

export const createRender = (body = []) => ({
  type: nodeTypes.FunctionDecl,
  id: createIdentifier('render'), // 默认名为 render
  params: [],
  body
})

export const createReturnStatement = (returnStatement) => ({
  type: nodeTypes.ReturnStatement,
  return: returnStatement // 应该返回 h() 函数调用，即 CallExpression
})

export const createCallExpr = (callee, args) => ({
  type: nodeTypes.CallExpression,
  callee: createIdentifier(callee),
  args
})
export const createArrayExpr = elements => ({
  type: nodeTypes.ArrayExpression,
  elements
})
export const createObjectLiteral = value => ({
  type: nodeTypes.ObjectLiteral,
  value
})
export const createStringLiteral = value => ({
  type: nodeTypes.StringLiteral,
  value
})
export const createIdentifier = (name) => ({
  type: nodeTypes.Identifier,
  name
})

export const createInterpolation = (content) => ({
  type: nodeTypes.Interpolation,
  content
})
