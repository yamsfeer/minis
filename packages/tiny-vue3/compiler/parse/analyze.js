import { tokenTypes } from './tokenize.js'

export const nodeTypes = {
  Root: Symbol('Root'),
  Element: Symbol('Element'),
  Text: Symbol('Text'),
}

const createRootNode = () => ({
  type: nodeTypes.Root,
  children: []
})

const createElementNode = (name) => ({
  type: nodeTypes.Element,
  tag: name,
  children: []
})

const createTextNode = (content) => ({
  type: nodeTypes.Text,
  content
})


export default function analyse(tokens) {
  const root = createRootNode()
  const stack = [root]

  while (tokens.length) {
    const parent = stack[stack.length - 1]
    const token = tokens.shift()

    switch (token.type) {
      case tokenTypes.tagOpen:
        const elementNode = createElementNode(token.name)

        parent.children.push(elementNode)
        stack.push(elementNode)
        break
      case tokenTypes.text:
        const textNode = createTextNode(token.content)
        parent.children.push(textNode)
        break
      case tokenTypes.tagEnd:
        stack.pop()
        break
    }
  }

  return root
}
