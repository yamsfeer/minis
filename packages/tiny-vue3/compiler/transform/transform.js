import {
  createRender,
  createCallExpr,
  createArrayExpr,
  createObjectLiteral,
  createStringLiteral,
  createIdentifier,
  createInterpolation,
  createReturnStatement,
} from './ast.js'
import { nodeTypes } from '../parse/parse.js'

const context = {
  currentNode: null,
  childIndex: 0,
  parent: null,
  replaceNode(newNode) {
    context.currentNode = newNode
    context.parent.children[context.childIndex] = newNode
  },
  removeNode() {
    if (!context.parent) return

    // 从父节点的 children 中删除
    context.parent.children.splice(context.childInex, 1)
    context.currentNode = null
  },
  nodeTransforms: [ // 转换节点
    transformRoot,
    transformElement,
    transformInterpolaton,
    transformText
  ]
}

function transformRoot(node) {
  if (node.type !== nodeTypes.Root)
    return

  return () => {
    /* Root 节点对应硬编码的 render 函数
       function render() { return ... } */
    const rtn = createReturnStatement(node.children[0].jsNode)
    const body = [rtn]
    node.jsNode = createRender(body)
  }
}

function transformElement(node) {
  if (node.type !== nodeTypes.Element)
    return

  return () => {
    /**
     * <div foo="bar"><p></p></div>
      * ===> h('div', { foo: 'bar' }, [h('p', null, [])])
     * const createCallExpr = (callee, args) => ({...})
     */

    const props = node.props.length
      ? node.props.reduce((acc, prop) => {
        acc[prop.name] = prop.value
        return acc
      }, {})
      : null
    node.jsNode = createCallExpr(
      'h',
      [
        createStringLiteral(node.tag), // 'div'
        createObjectLiteral(props), // props
        createArrayExpr(node.children.map(c => c.jsNode)) // children
      ]
    )
  }
}

function transformInterpolaton(node) {
  if (node.type !== nodeTypes.Interpolation)
    return

  /* {
    type: nodeTypes.Interpolation,
    content: {
      type: nodeTypes.Expression,
      content
    }
  } */
  node.jsNode = createInterpolation(node.content.content)
}

function transformText(node) {
  if (node.type !== nodeTypes.Text)
    return
  node.jsNode = createStringLiteral(node.content)
}

function traverse(root, context) {
  if (!root) return

  const exitFns = []

  context.currentNode = root
  for (const transform of context.nodeTransforms) {
    // transform 返回一个函数说明它需要在子节点处理完毕后进行
    const onExit = transform(root, context)
    onExit && exitFns.push(onExit)

    if (!context.currentNode) // 删除节点后，该节点的后序转换及其子树节点都不需要处理
      return
  }

  if (root.children) {
    root.children.forEach((child, index) => {
      context.parent = context.currentNode
      context.childIndex = index
      traverse(child, context)
    })
  }

  // 当前节点退出阶段，子节点已转换完成，注意从后往前遍历
  while (exitFns.length) {
    const fn = exitFns.pop()
    fn()
  }
}

export default function transform(ast) {
  traverse(ast, context)

  return ast
}
