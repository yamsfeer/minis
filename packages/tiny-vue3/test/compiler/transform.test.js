import transform from '../../compiler/transform/transform.js'
import { nodeTypes } from '../../compiler/parse/parse.js'

describe('compile:trasnform', () => {
  it('basic transform', () => {
    // <div><p>hello</p><p>world</p></div>
    const ast = {
      type: nodeTypes.Root,
      children: [{
        type: nodeTypes.Element,
        tag: 'div',
        props: [],
        children: [{
          type: nodeTypes.Element,
          tag: 'p',
          props: [{
            type: nodeTypes.Attributes,
            name: 'id',
            value: 'foo'
          }],
          children: [{ type: nodeTypes.Text, content: 'hello' }],
        }, {
          type: nodeTypes.Element,
          tag: 'p',
          props: [],
          children: [{ type: nodeTypes.Text, content: 'world' }]
        }]
      }]
    }

    const jsAST = transform(ast)
    expect(jsAST).toMatchSnapshot()
  })
})
