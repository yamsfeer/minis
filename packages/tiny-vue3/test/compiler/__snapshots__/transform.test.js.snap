// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compile:trasnform > basic transform 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "hello",
              "jsNode": {
                "type": Symbol(StringLiteral),
                "value": "hello",
              },
              "type": Symbol(Text),
            },
          ],
          "jsNode": {
            "args": [
              {
                "type": Symbol(StringLiteral),
                "value": "p",
              },
              {
                "type": Symbol(ObjectLiteral),
                "value": {
                  "id": "foo",
                },
              },
              {
                "elements": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "hello",
                  },
                ],
                "type": Symbol(ArrayExpression),
              },
            ],
            "callee": {
              "name": "h",
              "type": Symbol(Identifier),
            },
            "type": Symbol(CallExpression),
          },
          "props": [
            {
              "name": "id",
              "type": Symbol(Attributes),
              "value": "foo",
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
        {
          "children": [
            {
              "content": "world",
              "jsNode": {
                "type": Symbol(StringLiteral),
                "value": "world",
              },
              "type": Symbol(Text),
            },
          ],
          "jsNode": {
            "args": [
              {
                "type": Symbol(StringLiteral),
                "value": "p",
              },
              {
                "type": Symbol(ObjectLiteral),
                "value": null,
              },
              {
                "elements": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "world",
                  },
                ],
                "type": Symbol(ArrayExpression),
              },
            ],
            "callee": {
              "name": "h",
              "type": Symbol(Identifier),
            },
            "type": Symbol(CallExpression),
          },
          "props": [],
          "tag": "p",
          "type": Symbol(Element),
        },
      ],
      "jsNode": {
        "args": [
          {
            "type": Symbol(StringLiteral),
            "value": "div",
          },
          {
            "type": Symbol(ObjectLiteral),
            "value": null,
          },
          {
            "elements": [
              {
                "args": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "p",
                  },
                  {
                    "type": Symbol(ObjectLiteral),
                    "value": {
                      "id": "foo",
                    },
                  },
                  {
                    "elements": [
                      {
                        "type": Symbol(StringLiteral),
                        "value": "hello",
                      },
                    ],
                    "type": Symbol(ArrayExpression),
                  },
                ],
                "callee": {
                  "name": "h",
                  "type": Symbol(Identifier),
                },
                "type": Symbol(CallExpression),
              },
              {
                "args": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "p",
                  },
                  {
                    "type": Symbol(ObjectLiteral),
                    "value": null,
                  },
                  {
                    "elements": [
                      {
                        "type": Symbol(StringLiteral),
                        "value": "world",
                      },
                    ],
                    "type": Symbol(ArrayExpression),
                  },
                ],
                "callee": {
                  "name": "h",
                  "type": Symbol(Identifier),
                },
                "type": Symbol(CallExpression),
              },
            ],
            "type": Symbol(ArrayExpression),
          },
        ],
        "callee": {
          "name": "h",
          "type": Symbol(Identifier),
        },
        "type": Symbol(CallExpression),
      },
      "props": [],
      "tag": "div",
      "type": Symbol(Element),
    },
  ],
  "jsNode": {
    "body": [
      {
        "return": {
          "args": [
            {
              "type": Symbol(StringLiteral),
              "value": "div",
            },
            {
              "type": Symbol(ObjectLiteral),
              "value": null,
            },
            {
              "elements": [
                {
                  "args": [
                    {
                      "type": Symbol(StringLiteral),
                      "value": "p",
                    },
                    {
                      "type": Symbol(ObjectLiteral),
                      "value": {
                        "id": "foo",
                      },
                    },
                    {
                      "elements": [
                        {
                          "type": Symbol(StringLiteral),
                          "value": "hello",
                        },
                      ],
                      "type": Symbol(ArrayExpression),
                    },
                  ],
                  "callee": {
                    "name": "h",
                    "type": Symbol(Identifier),
                  },
                  "type": Symbol(CallExpression),
                },
                {
                  "args": [
                    {
                      "type": Symbol(StringLiteral),
                      "value": "p",
                    },
                    {
                      "type": Symbol(ObjectLiteral),
                      "value": null,
                    },
                    {
                      "elements": [
                        {
                          "type": Symbol(StringLiteral),
                          "value": "world",
                        },
                      ],
                      "type": Symbol(ArrayExpression),
                    },
                  ],
                  "callee": {
                    "name": "h",
                    "type": Symbol(Identifier),
                  },
                  "type": Symbol(CallExpression),
                },
              ],
              "type": Symbol(ArrayExpression),
            },
          ],
          "callee": {
            "name": "h",
            "type": Symbol(Identifier),
          },
          "type": Symbol(CallExpression),
        },
        "type": Symbol(ReturnStatement),
      },
    ],
    "id": {
      "name": "render",
      "type": Symbol(Identifier),
    },
    "params": [],
    "type": Symbol(FunctionDecl),
  },
  "type": Symbol(Root),
}
`;

exports[`compile:trasnform basic transform 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "content": "hello",
              "jsNode": {
                "type": Symbol(StringLiteral),
                "value": "hello",
              },
              "type": Symbol(Text),
            },
          ],
          "jsNode": {
            "args": [
              {
                "type": Symbol(StringLiteral),
                "value": "p",
              },
              {
                "type": Symbol(ObjectLiteral),
                "value": {
                  "id": "foo",
                },
              },
              {
                "elements": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "hello",
                  },
                ],
                "type": Symbol(ArrayExpression),
              },
            ],
            "callee": {
              "name": "h",
              "type": Symbol(Identifier),
            },
            "type": Symbol(CallExpression),
          },
          "props": [
            {
              "name": "id",
              "type": Symbol(Attributes),
              "value": "foo",
            },
          ],
          "tag": "p",
          "type": Symbol(Element),
        },
        {
          "children": [
            {
              "content": "world",
              "jsNode": {
                "type": Symbol(StringLiteral),
                "value": "world",
              },
              "type": Symbol(Text),
            },
          ],
          "jsNode": {
            "args": [
              {
                "type": Symbol(StringLiteral),
                "value": "p",
              },
              {
                "type": Symbol(ObjectLiteral),
                "value": null,
              },
              {
                "elements": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "world",
                  },
                ],
                "type": Symbol(ArrayExpression),
              },
            ],
            "callee": {
              "name": "h",
              "type": Symbol(Identifier),
            },
            "type": Symbol(CallExpression),
          },
          "props": [],
          "tag": "p",
          "type": Symbol(Element),
        },
      ],
      "jsNode": {
        "args": [
          {
            "type": Symbol(StringLiteral),
            "value": "div",
          },
          {
            "type": Symbol(ObjectLiteral),
            "value": null,
          },
          {
            "elements": [
              {
                "args": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "p",
                  },
                  {
                    "type": Symbol(ObjectLiteral),
                    "value": {
                      "id": "foo",
                    },
                  },
                  {
                    "elements": [
                      {
                        "type": Symbol(StringLiteral),
                        "value": "hello",
                      },
                    ],
                    "type": Symbol(ArrayExpression),
                  },
                ],
                "callee": {
                  "name": "h",
                  "type": Symbol(Identifier),
                },
                "type": Symbol(CallExpression),
              },
              {
                "args": [
                  {
                    "type": Symbol(StringLiteral),
                    "value": "p",
                  },
                  {
                    "type": Symbol(ObjectLiteral),
                    "value": null,
                  },
                  {
                    "elements": [
                      {
                        "type": Symbol(StringLiteral),
                        "value": "world",
                      },
                    ],
                    "type": Symbol(ArrayExpression),
                  },
                ],
                "callee": {
                  "name": "h",
                  "type": Symbol(Identifier),
                },
                "type": Symbol(CallExpression),
              },
            ],
            "type": Symbol(ArrayExpression),
          },
        ],
        "callee": {
          "name": "h",
          "type": Symbol(Identifier),
        },
        "type": Symbol(CallExpression),
      },
      "props": [],
      "tag": "div",
      "type": Symbol(Element),
    },
  ],
  "jsNode": {
    "body": [
      {
        "return": {
          "args": [
            {
              "type": Symbol(StringLiteral),
              "value": "div",
            },
            {
              "type": Symbol(ObjectLiteral),
              "value": null,
            },
            {
              "elements": [
                {
                  "args": [
                    {
                      "type": Symbol(StringLiteral),
                      "value": "p",
                    },
                    {
                      "type": Symbol(ObjectLiteral),
                      "value": {
                        "id": "foo",
                      },
                    },
                    {
                      "elements": [
                        {
                          "type": Symbol(StringLiteral),
                          "value": "hello",
                        },
                      ],
                      "type": Symbol(ArrayExpression),
                    },
                  ],
                  "callee": {
                    "name": "h",
                    "type": Symbol(Identifier),
                  },
                  "type": Symbol(CallExpression),
                },
                {
                  "args": [
                    {
                      "type": Symbol(StringLiteral),
                      "value": "p",
                    },
                    {
                      "type": Symbol(ObjectLiteral),
                      "value": null,
                    },
                    {
                      "elements": [
                        {
                          "type": Symbol(StringLiteral),
                          "value": "world",
                        },
                      ],
                      "type": Symbol(ArrayExpression),
                    },
                  ],
                  "callee": {
                    "name": "h",
                    "type": Symbol(Identifier),
                  },
                  "type": Symbol(CallExpression),
                },
              ],
              "type": Symbol(ArrayExpression),
            },
          ],
          "callee": {
            "name": "h",
            "type": Symbol(Identifier),
          },
          "type": Symbol(CallExpression),
        },
        "type": Symbol(ReturnStatement),
      },
    ],
    "id": {
      "name": "render",
      "type": Symbol(Identifier),
    },
    "params": [],
    "type": Symbol(FunctionDecl),
  },
  "type": Symbol(Root),
}
`;
